### PR TITLE
feat: find duplicate objects init

### LINKED ISSUE #3362: Introduce duplicate query for objects
On a record's detail page, we would like to display to user if there are any record that appears to be a duplicate.

 
## Backend

For v1 we could start with those 3 rules:
```
const duplicateCriterias = 
[
    {
        "object": "company",
        "fields": ["domain"]
    },
    {
        "type": "person",
        "fields":  ["firstName", "lastName"], 
    },
    {
        "type": "person",
        "duplicateCriteria":  ["email"], 
    }
];
```
In the future this will be configurable for any type of object as the config will be pulled from DB and configured individually for each workspace.  Also we will likely add other columns such as `actionOnCreate : disallow/alert`, `actionOnEdit: alert/no_regression` or `alertText: string`.

And the API you will need to create to retrieve duplicates on the frontend for a given record should look like something like this:

Screenshot 2024-02-08 at 17 05 47

Note there are two parameters: `data` and `id`, they can be null but both can't be null at the same time.

If data is given then we compute the duplicate criteria based on the data (e.g. we can pass {"domain": "airbnb.com", "name":"Airbnb"}.
If ID is given, then we pull the data  for that given record from DB.
In both cases, we will then find if there's any duplicate by applying the `duplicateCriterias` and return the list of duplicates. We just return an empty array if there isn't any, no need for a specific config. 

This should result in the creation of as many queries as there are non-technical objects: `companyDuplicates, personDuplicates, opportunityDuplicates, customObject1Duplicate, customObject2Duplicates`. Every object that is visible in /settings/object page should have its dedicated query. Technical objects such as "View" or "ViewFilter" should not have a query.

**Before jumping into this issue, please note that our backend is very complex and different from traditional infrastructure because of the specific requirements of a CRM** (custom objects/fields). We've got one schema per workspace in Postgres, each workspace can define the metadata for their business and our engine automatically generates the sql migration / the GraphQL queries based on that. So if you're new to this project it would probably take a few hours for you to understand this.

Test coverage is poor for backend but please write at least a few basic tests for this.

## Frontend

If there is no duplicate, we shouldn't display anything.

If there is a duplicate, we should show it in a similar fashion to how relationships are currently displayed:
Screenshot 2024-02-08 at 16 59 23

### PR DESCRIPTION
**Pull request by rostaklein**
### Query via ID:
Screenshot 2024-02-19 at 22 13 29

### Query via partial data:
Screenshot 2024-02-19 at 22 13 44

### Duplicates displayed in the UI (no actions yet):
image

### CODE REVIEW

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts
Comment (magrinj) [inline packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts:19]: I know in the ticket the keys are `object` and `fields` but I rather name them `objectName` and `fieldNames` to make to object more clear :)

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts
Comment (magrinj) [inline packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts:17]:
```suggestion
type RecordDuplicateCriteria = {
```

File: packages/twenty-server/src/workspace/workspace-resolver-builder/workspace-resolver.factory.ts
Comment (magrinj) [inline packages/twenty-server/src/workspace/workspace-resolver-builder/workspace-resolver.factory.ts:54]: @rostaklein Maybe we should add a condition bellow to only add query of object within the `duplicateCriteria` array.
You can by example in the `for...of` loop skip for the method name `findDuplicates` if the objectName is not matching the object metadata singular name.

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts
Comment (magrinj) [inline packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts:24]: @rostaklein I think this should be moved outside as we'll also use this array in the `workspace-resolver.factory.ts`.
You can create into `workspace-resolver-builder` a folder `constants`

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts
Comment (magrinj) [reply packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts:17]: BTW: As we're going to move the duplicate criteria outside we should move this type into the `interfaces` folder and instead use `interface` over `type`.

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts
Comment (magrinj) [reply packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts:24]: BTW: In English the plural of `criteria` is `criteria`, maybe we can name this `duplicateCriteriaCollection` ?
We got the same problem with `metadata` haha 😂

**Review by magrinj — COMMENTED** [review]
Looks really promising, thanks a lot for this awesome work !! 🎉😍

File: packages/twenty-server/src/workspace/workspace-schema-builder/utils/get-resolver-args.util.ts
Comment (rostaklein) [inline packages/twenty-server/src/workspace/workspace-schema-builder/utils/get-resolver-args.util.ts:79]: This type is not ideal. I dont want to use `InputTypeDefinitionKind.Filter` though, because we wont allow specifying filter types. Am I right? What would be your suggestion to use here? 🤔

File: packages/twenty-server/src/workspace/workspace-schema-builder/utils/should-skip-query.util.ts
Comment (rostaklein) [inline packages/twenty-server/src/workspace/workspace-schema-builder/utils/should-skip-query.util.ts:22]: Not sure if this is the right placement of such a file, but its necessary for both the type and resolver factories.

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts
Comment (rostaklein) [inline packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts:19]: I could probably easily create some unit tests for this factory to have "at least some" coverage. Best one would be actually a proper integration test that spins up the whole app, has a db behind and the test actually calls the graphql query. That would require much more effort to bootstrap though.

File: packages/twenty-front/src/modules/object-record/hooks/useFindDuplicateRecords.ts
Comment (rostaklein) [inline packages/twenty-front/src/modules/object-record/hooks/useFindDuplicateRecords.ts:14]: There is no pagination now. We might think of introducing it later. I dont think that showing more than just a few duplicates makes sense for the current use case though, so i didnt bother

File: packages/twenty-server/src/workspace/workspace-schema-builder/utils/get-resolver-args.util.ts
Comment (magrinj) [reply packages/twenty-server/src/workspace/workspace-schema-builder/utils/get-resolver-args.util.ts:79]: I totally agree with you on that.
Also this `data` input should be limited to the provided `criteriaCollection` not all the fields of the object.
Only providing the fields exposed by `criteriaCollection` is adding more complexity because we need to create a new graphQL input type based on the `criteriaCollection`, and this one is using `columnNames` and not `fieldNames`.
Do we keep it that way for now @charlesBochet ?

**Review by magrinj — DISMISSED** [review]
Thanks a lot for you work, it's really good ! 🙂
Some minor changes, and discussions are blocking the PR for now as we need to take some decisions in the core team where I've ping some of them.

File: packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts
Comment (magrinj) [inline packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts:20]: BTW: The array contains more criteria than the ticket, is it just for testing ?

Comment (rostaklein) [reply packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts:20]: Yes that was just for the testing purpose :) At the same time i though this might make sense (if i got a duplicate email, then its probably the same person?). Im ok with deleting it though as it was not in the original ticket.

File: packages/twenty-server/src/workspace/workspace-schema-builder/utils/get-resolver-args.util.ts
Comment (FelixMalfait) [reply packages/twenty-server/src/workspace/workspace-schema-builder/utils/get-resolver-args.util.ts:79]: "Also this data input should be limited to the provided criteriaCollection not all the fields of the object."
Not sure I understand/agree with that? The de-duplication logic is server-side, it's not the frontend's role to stay up-to-date on which field or not might trigger a duplicate alert (it should be allowed to send everything they want on that object), that logic will evolve and be different for each workspace long-term.

File: packages/twenty-front/src/modules/object-record/record-duplicates-card/components/RecordDuplicatesFieldCardSection.tsx
Comment (FelixMalfait) [reply packages/twenty-front/src/modules/object-record/record-duplicates-card/components/RecordDuplicatesFieldCardSection.tsx:85]: For now it should be displayed the same way we display other relationships. 

Ex on people:
Screenshot 2024-02-20 at 14 02 35

File: packages/twenty-server/src/workspace/workspace-schema-builder/utils/get-resolver-args.util.ts
Comment (FelixMalfait) [reply packages/twenty-server/src/workspace/workspace-schema-builder/utils/get-resolver-args.util.ts:79]: I agree with `columnNames` rather than `fieldNames`!

File: packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts
Comment (FelixMalfait) [reply packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts:3]: I agree with columnNames rather than fieldNames!

File: packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts
Comment (FelixMalfait) [reply packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts:6]: Sorry about providing that example which was incorrect. The correct field is `domainName`

**Comment by FelixMalfait** [comment]
Great work @rostaklein 🙌 🙌 🙌

File: packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts
Comment (FelixMalfait) [reply packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts:20]: https://github.com/twentyhq/twenty/issues/4048#issuecomment-1953135739 Made me think that maybe we can add person.LinkedinLinkUrl as a default too?

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts
Comment (rostaklein) [reply packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts:30]: Makes sense :) I though `parseResult` would parse the composite fields such as `nameFirstName` to `name { firstName }`, but it doesnt, so all good here.

File: packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts
Comment (rostaklein) [reply packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts:20]: Good point :) Ive added `person.linkedinLinkUrl` as well.

File: packages/twenty-front/src/modules/object-record/record-duplicates-card/components/RecordDuplicatesFieldCardSection.tsx
Comment (rostaklein) [reply packages/twenty-front/src/modules/object-record/record-duplicates-card/components/RecordDuplicatesFieldCardSection.tsx:85]: Theres one more problem on FE. For objects that the query is not generated it throws an error (kinda expected), bcos the generated query doesnt exist for such object. What should we do in such case? How could we prevent this other than knowing upfront that a certain object does not have a duplicate criteria? 🤔

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts
Comment (rostaklein) [reply packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts]: Ive tried to improve the readability a bit, but it is harder than it seems 😄 Im open to further suggestions. The other part of this comment regarding an empty array instead of all the results - damn, how do i create a query that resolves zero records without any hacks? 🤔 😆 Cant figure it out.

File: packages/twenty-front/src/modules/object-record/record-duplicates-card/components/RecordDuplicatesFieldCardSection.tsx
Comment (magrinj) [reply packages/twenty-front/src/modules/object-record/record-duplicates-card/components/RecordDuplicatesFieldCardSection.tsx:85]: Hmm interesting one 😅
Maybe we can move the criteria collection in the core and add something to query them ?
This way the front-end knows which ones has duplicates ?
Or we rollback to create duplicates for all objects and return something empty on the ones that doesn't implement criteria ?
cc @charlesBochet

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts
Comment (magrinj) [reply packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts]: Actually parseResult is parsing stuff like `nameFirstName` but they're transformed only if they start with triple underscore `___` (it's a small convention that we put in place 😅)

File: packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts
Comment (magrinj) [reply packages/twenty-server/src/workspace/workspace-resolver-builder/constants/duplicate-criteria.constants.ts:9]: @rostaklein Ok we'll go that way in this PR, but I'll create a ticket to use `fieldNames` instead of `columnNames` and do the compute in the factory to pre-process the `columnNames` based on the `fieldNames`.
This PR is big enough, so let's split that haha !

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts
Comment (magrinj) [inline packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts:116]:
```suggestion
        criteria.columnNames.map((columnName) => [
```

Better not to use abbreviation :)

File: packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts
Comment (magrinj) [inline packages/twenty-server/src/workspace/workspace-query-builder/factories/find-duplicates-query.factory.ts:137]:
```suggestion
      (duplicateCriteria) => duplicateCriteria === objectMetadataItem.nameSingular,
```

File: packages/twenty-server/src/workspace/workspace-schema-builder/utils/get-resolver-args.util.ts
Comment (magrinj) [reply packages/twenty-server/src/workspace/workspace-schema-builder/utils/get-resolver-args.util.ts:79]: @rostaklein Ok we'll keep it that way for now :)

File: packages/twenty-front/src/modules/object-record/record-duplicates-card/components/RecordDuplicatesFieldCardSection.tsx
Comment (magrinj) [reply packages/twenty-front/src/modules/object-record/record-duplicates-card/components/RecordDuplicatesFieldCardSection.tsx:85]: Ok thanks for the clarification, so @rostaklein I'm sorry I've asked to only expose resolver based on the criteria collection, but we'll expose a resolve for all objects even the ones without criteria, and it should return an empty array for them.

**Comment by rostaklein** [comment]
Ive addressed another round of comments :) Frontend part is now fixed as well & screens in the description updated.

Basic unit tests coverage added, but to be honest, i dont think it tests much. The e2e (integration) test would be much more useful in this case, but if i am looking at them correctly, they are unmaintained now and I am not able to run them locally. Probably a thing to tackle in another issue.

**Review by charlesBochet — APPROVED** [review]
@rostaklein Thanks for the work, I've made some small refactoring on the frontend.
Also, I think we should use InputTypeDefinitionKind.Update as our data input for findDuplicates ; the question we ask this enpoint is "what if I create the following record, would it be a duplicate", so it makes sense to pass the same payload as create mutation endpoint

**Comment by charlesBochet** [comment]
@rostaklein we are good to go, thank you!

**Comment by rostaklein** [comment]
> InputTypeDefinitionKind.Update

I think you meant `InputTypeDefinitionKind.Create`, right? :) Anyways, thank you for the final commits.

**Comment by charlesBochet** [comment]
Yes `InputTypeDefinitionKind.Create`, right!

**Pull request merged**
